File: Programmers/Level 2/11.cpp
#include <string>
#include <vector>

using namespace std;

int solution(string skill, vector<string> skill_trees) {
    int answer = 0;
    for (int i = 0; i < skill_trees.size(); i++) {
        bool flag = false;
        int now = 0;
        for (int j = 0; j < skill.size(); j++) {
            for (int k = 0; k < skill_trees[i].size(); k++) {
                if (skill[j] == skill_trees[i][k]) {
                    //같은 스킬 발견
                    if (k < now) {
                        //선행 스킬을 무시한 경우
                        flag = true;
                    }
                    else {
                        now = k;//현재 인덱스 저장
                    }
                    break;
                }
                else {
                    //스킬 발견하지 못했는데 스킬 트리 끝까지 갔으면
                    if (k == skill_trees[i].size() - 1) {
                        now = skill_trees[i].size();
                    }
                }
            }
            if (flag)
                break;
        }
        if (!flag)
            answer++;
    }
    
    return answer;
}

File: Programmers/Level 3/11.py
answer = ["ICN"]
visited = [False] * 11000

def dfs(now, cnt, tickets):
    tmp = 0
    if cnt == len(tickets):
        return True 
    
    for i in range(len(tickets)):
        if tickets[i][0] == now and visited[i] == False:
            tmp += 1
            visited[i] = True            
            answer.append(tickets[i][1])
            flag = dfs(tickets[i][1], cnt+1, tickets)
            if flag == False:
                tmp = 0
                visited[i] = False
                answer.pop()

    if tmp == 0 and cnt != len(tickets):
        return False
    
def solution(tickets):
    global answer
    tickets.sort(key=lambda x:x[1]) # 알파벳 순 정렬
    dfs("ICN", 0, tickets)
    return answer
